src/org/opensolaris/opengrok/search/context/Context.java

```diff
@@ -64,34 +73,164 @@
+        boolean limit, int tabSize) {
+
+        if (isEmpty()) {
+            return false;
+        }
+
+        Document doc;
+        try {
+            doc = searcher.doc(docId);
+        } catch (IOException e) {
+            LOGGER.log(Level.WARNING, "ERROR getting searcher doc(int)", e);
+            return false;
+        }
+
+        Definitions tags = null;
+        try {
+            IndexableField tagsField = doc.getField(QueryBuilder.TAGS);
+            if (tagsField != null) {
+                tags = Definitions.deserialize(tagsField.binaryValue().bytes);
+            }
+        } catch (ClassNotFoundException|IOException e) {
+            LOGGER.log(Level.WARNING, "ERROR Definitions.deserialize(...)", e);
+            return false;
+        }
+
+        Scopes scopes;
+        try {
+            IndexableField scopesField = doc.getField(QueryBuilder.SCOPES);
+            if (scopesField != null) {
+                scopes = Scopes.deserialize(scopesField.binaryValue().bytes);
+            } else {
+                scopes = new Scopes();
+            }
+        } catch (ClassNotFoundException|IOException e) {
+            LOGGER.log(Level.WARNING, "ERROR Scopes.deserialize(...)", e);
+            return false;
+        }
+
+        /*
+         * UnifiedHighlighter demands an analyzer "even if in some
+         * circumstances it isn't used"; here it is not meant to be used.
+         */
+        PlainAnalyzerFactory fac = PlainAnalyzerFactory.DEFAULT_INSTANCE;
+        FileAnalyzer anz = fac.getAnalyzer();
+
+        String path = doc.get(QueryBuilder.PATH);
+        String pathE = Util.URIEncodePath(path);
+        String urlPrefixE = urlPrefix == null ? "" : Util.URIEncodePath(
+            urlPrefix);
+        String moreURL = morePrefix == null ? null : Util.URIEncodePath(
+            morePrefix) + pathE + "?" + queryAsURI;
+
+        ContextArgs args = new ContextArgs(env.getContextSurround(),
+            env.getContextLimit());
+        /**
+         * Lucene adds to the following value in FieldHighlighter, so avoid
+         * integer overflow by not using Integer.MAX_VALUE -- Short is good
+         * enough.
+         */
+        int linelimit = limit ? args.getContextLimit() : Short.MAX_VALUE;
+
+        ContextFormatter formatter = new ContextFormatter(args);
+        formatter.setUrl(urlPrefixE + pathE);
+        formatter.setDefs(tags);
+        formatter.setScopes(scopes);
+        formatter.setMoreUrl(moreURL);
+        formatter.setMoreLimit(linelimit);
+
+        OGKUnifiedHighlighter uhi = new OGKUnifiedHighlighter(env,
+            searcher, anz);
+        uhi.setBreakIterator(() -> new StrictLineBreakIterator());
+        uhi.setFormatter(formatter);
+        uhi.setTabSize(tabSize);
+
+        try {
+            List<String> fieldList = qbuilder.getContextFields();
+            String[] fields = new String[fieldList.size()];
+            fieldList.toArray(fields);
```

Comment:
Could you please consider using this code: `String[] fields = fieldList.toArray(new String[fieldList.size()]);` ? I think this might be a little dangerous since it relies heavily on the size of the `fields` array. If someone would change it then it would break.

Reply:
ok revised